### DISCUSSION TITLE
Understanding hub orientation (balance bot)

### DISCUSSION THREAD

**Discussion by coder-ella** [pybricks/support#1504] [Q&A]
Hi I am trying to understand why the [BalanceBot](https://www.patreon.com/posts/balance-on-ball-94063633) axis are set to Top = 1,1,0 and Front = -1,1,0

In this diagram T is top and F is front (at least how I am thinking about it)
![image](https://github.com/pybricks/support/assets/66184276/548cbfcf-60ca-4ad8-b10c-17c34c70a60f)

![image](https://github.com/pybricks/support/assets/66184276/9617f989-f8c9-4a1d-bd5c-343de75e92ba)

This is the documentation diagram:  (https://docs.pybricks.com/en/stable/signaltypes.html#reference-frames)
![image](https://github.com/pybricks/support/assets/66184276/637d125c-19bd-4091-9f27-0520549e88be)

Based on the above reference-frame documentation, I have drawn the following diagram, where, in the diagram:  
1 represents my understanding of the normal robot axis. (Top = (0,0,1)  and front = (1,0,0))
2 represents what I think the Top and front for the robot in balance bot should be.  (Top = (0,1,0) and Front = (0,0,1))
And 3 represents what I think the code represents as the front and top (Top = 1,1,0 and Front = -1,1,0). Something that is confusing to me is how are Top and Front are in the same plane (XY). 

![image](https://github.com/pybricks/support/assets/66184276/f28fac4c-6399-4140-a8a7-489543f15451)

ps: I am in 6th grade! and I wrote this question with help from my dad.

**Comment by laurensvalk** [comment]
Great question, and great work making those diagrams! To help answer it, perhaps it helps if I add how I arrived at these values.

I had previously written code for less complicated [two-wheeled balancing robots](https://www.youtube.com/watch?v=muVCJEmVa_M). 

My thinking was that I could take two "copies" of that code for a ball balancer. One for its X rotation driving one wheel, and one for its Y rotation driving the other wheel. And hope that they would not influence each other much. This is a fairly reasonable assumption as long as the robot stays almost perfectly upright. (But if it isn't, all hope is lost anyway :smile: )

To implement this idea, I'd need the gyro values for the axes corresponding to the rotation of those motors.

A somewhat trivial way to accomplish that would be to physically mount it such that the X and Y would just align exactly with the motors. But for aesthetic reasons, I wanted to mount it diagonally. And so you're left with the challenge you're currently looking into. That is, how to tell the robot how the hub is mounted, such that it knows to give the proper gyro values for balancing.

So the goal here was to get the _wheel axes_ aligned with the X and Y axes you drew. And when you do that, the hub is would sit diagonally in your diagrams, with the arrow that comes out of the side with the display, lying in the XY plane along (1, 1, 0).

Let me try to continue this post tomorrow with a proper diagram in the context of the robot, but maybe this will help you start wrap your head around it intuitively.

**Reply by laurensvalk** [reply]
Hey @coder-ella - hope your project is going well! Here is a diagram to further illustrate the above:

Here you can see the motors line up with the red and green axes. And the hub being "diagonal" as a result. The hub faces then point in the direction of the orange and pink axes, which are the values you've found in the code.

![image](https://github.com/pybricks/support/assets/12326241/4307e0e8-eff0-4b05-8127-d6da3f9586e0)
